File: src/lane_detection/src/lane_detection_node.cpp
#include <cv_bridge/cv_bridge.h>

#include <chrono>
#include <memory>
#include <opencv2/opencv.hpp>
#include <opencv2/videoio.hpp>
#include <rclcpp/logging.hpp>
#include <rclcpp/node.hpp>
#include <rclcpp/qos.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/subscription_base.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <std_msgs/msg/detail/float64__struct.hpp>
#include <std_msgs/msg/float64.hpp>
#include <std_msgs/msg/int32.hpp>
#include <utility>

#include "../include/lane_detection/lane_detector.h"
/// custom namespace for readability
/// /////////////////////////////////////////////////////////
namespace lane_detection_node_ns {
// rename types for readability
// rename image
using img_msg = sensor_msgs::msg::Image;
using img_ptr = img_msg::SharedPtr;
using img_pub = rclcpp::Publisher<img_msg>::SharedPtr;
using img_sub = rclcpp::Subscription<img_msg>::SharedPtr;

// rename int
using int_msg = std_msgs::msg::Int32;
using int_pub = rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr;

// rename double
using double_msg = std_msgs::msg::Float64;
using double_pub = rclcpp::Publisher<double_msg>::SharedPtr;
// end of renaming

// adding ros2
using cv_bridge::CvImage;
using rclcpp::init;
using rclcpp::SensorDataQoS;
using rclcpp::shutdown;
using rclcpp::spin;
using Header = std_msgs::msg::Header_<std::allocator<void>>;
using timer_ptr = rclcpp::TimerBase::SharedPtr;

// opencv
using cv::Mat;
using cv::Point;
using cv::Point2f;

// adding std
using std::bind;
using std::cout;
using std::endl;
using std::string;
using std::vector;
using namespace std::placeholders;
using std::make_shared;
using std::chrono::milliseconds;

// adding lane detector namespace
using ld_ns::Lane_detector;
};  // namespace lane_detection_node_ns
using namespace lane_detection_node_ns;

/// Class Node Declaration
/// /////////////////////////////////////////////////////////
class lane_detection_node : public rclcpp::Node {
   public:
    lane_detection_node() : Node("lane_detection_node") {
        // contruct lane_detector class and pass this node as a arg
        ld_ = std::make_unique<Lane_detector>(this);

        // Declare parameters for the camera
        this->declare_parameter("camera_index", 0);  // index of the camera
        this->declare_parameter("width", 640);       // width of the cam feed
        this->declare_parameter("height", 480);      // height of the cam feed
        this->declare_parameter("fps", 60);          // fps of the cam feed

        // Get parameters for camera
        int cam_index = this->get_parameter("camera_index").as_int();
        int width = this->get_parameter("width").as_int();
        int height = this->get_parameter("height").as_int();
        int fps = this->get_parameter("fps").as_int();

        // Create the working pipeline
        string pipeline = create_pipeline(cam_index, width, height, fps);

        cap_ = cv::VideoCapture(pipeline, cv::CAP_GSTREAMER);

        if (!cap_.isOpened()) {
            RCLCPP_ERROR(this->get_logger(), "Failed to open camera with pipeline: %s",
                         pipeline.c_str());
            return;
        }

        // setting up publishers
        raw_video_pub_ = this->create_publisher<img_msg>("camera/raw_video", 1);
        result_img_pub_ =
            this->create_publisher<img_msg>("camera/lane_detection_result", 1);
        bev_img_pub_ = this->create_publisher<img_msg>("camera/bev", 1);
        mask_img_pub_ = this->create_publisher<img_msg>("camera/bin_mask", 1);
        filter_img_pub_ = this->create_publisher<img_msg>("camera/filter_bin_mask", 1);
        delta_pub_ = this->create_publisher<int_msg>("lane_detect/delta", 1);
        position_confidence_pub_ =
            this->create_publisher<double_msg>("lane_detect/position_confidence", 1);
        symmetry_confidence_pub_ =
            this->create_publisher<double_msg>("lane_detect/symmetry_confidence", 1);

        RCLCPP_INFO(this->get_logger(), "lane_detection_node has started");

        timer_ = this->create_wall_timer(
            milliseconds(int(1 / fps)), bind(&lane_detection_node::process_frame, this));
    }

   private:
    // publisher objects
    img_pub raw_video_pub_;
    img_pub mask_img_pub_;
    img_pub filter_img_pub_;
    img_pub result_img_pub_;
    img_pub bev_img_pub_;
    int_pub delta_pub_;
    double_pub position_confidence_pub_;
    double_pub symmetry_confidence_pub_;

    // lane detector object
    std::unique_ptr<Lane_detector> ld_;

    // videoCapture object
    cv::VideoCapture cap_;

    // callback timer
    timer_ptr timer_;

    void pub_img(img_pub pub, Mat img, string encoding) {
        img_ptr temp_ptr = CvImage(Header(), encoding, img).toImageMsg();
        temp_ptr->header.stamp = this->get_clock()->now();
        temp_ptr->header.frame_id = "video_feed";

        // publish  result
        pub->publish(*temp_ptr);
    }

    void process_frame() {
        int_msg delta;
        double_msg position_confidence;
        double_msg symmetry_confidence;
        Mat frame;

        // get frame and process it
        cap_ >> frame;

        if (!frame.empty()) {
            if (frame.channels() == 4) {
                // need to convert the BGRx to BGR
                cv::cvtColor(frame, frame, cv::COLOR_BGRA2BGR);
            }
            auto results = ld_->step(frame);

            delta.data = results.delta;
            RCLCPP_DEBUG(this->get_logger(), "delta: %d", delta.data);

            // get confidence values
            position_confidence.data = results.position_conf;
            symmetry_confidence.data = results.symmetry_conf;

            // pub servo value and confidence values
            delta_pub_->publish(delta);
            position_confidence_pub_->publish(position_confidence);
            symmetry_confidence_pub_->publish(symmetry_confidence);

            // pub images for lane detection algorithm
            RCLCPP_DEBUG(this->get_logger(), "pub images");
            pub_img(raw_video_pub_, frame, "bgr8");
            pub_img(result_img_pub_, results.ld_img, "bgr8");
            pub_img(bev_img_pub_, results.bev, "bgr8");
            pub_img(mask_img_pub_, results.bin_mask, "mono8");
            pub_img(filter_img_pub_, results.filter_mask, "mono8");
        }
    }

    string create_pipeline(int cam_index, int width, int height, int fps) {
        std::ostringstream pipeline;

        pipeline << "v4l2src device=/dev/video" << cam_index
                 << " ! video/x-raw,width=(int)" << width << ",height=(int)"
                 << height << ",framerate=(fraction)" << fps << "/1 "
                 << "! nvvidconv"
                 << "! video/x-raw,format=BGRx "  // Specify output format
                 << "! videoconvert "             // Ensure proper conversion
                 << "! video/x-raw,format=BGR "   // Convert to BGR for OpenCV
                 << "! appsink sync=false";

        RCLCPP_INFO(this->get_logger(), "Pipeline: %s", pipeline.str().c_str());
        return pipeline.str();
    }
};

/// Main ////////////////////////////////////////////////////
int main(int argc, char **argv) {
    init(argc, argv);
    auto node = make_shared<lane_detection_node>();
    spin(node);
    shutdown();
    return 0;
}

File: src/lane_detection_py/lane_detection_py/lane_detection_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from std_msgs.msg import Int32, Float64
from cv_bridge import CvBridge
import cv2
from .lane_detector import LaneDetector

class LaneDetectionNode(Node):
    def __init__(self):
        super().__init__('lane_detection_node')
        
        # Initialize your custom lane detector logic
        self.ld = LaneDetector(self) 

        # 1. Declare and Get Parameters
        self.declare_parameter('camera_index', 0)
        self.declare_parameter('width', 640)
        self.declare_parameter('height', 480)
        self.declare_parameter('fps', 60)

        cam_index = self.get_parameter('camera_index').value
        width = self.get_parameter('width').value
        height = self.get_parameter('height').value
        fps = self.get_parameter('fps').value

        # 2. Setup OpenCV to ROS Image Bridge
        self.bridge = CvBridge()

        # 3. Create and Open GStreamer Pipeline
        pipeline = self.create_pipeline(cam_index, width, height, fps)
        self.cap = cv2.VideoCapture(pipeline, cv2.CAP_GSTREAMER)

        if not self.cap.isOpened():
            self.get_logger().error(f'Failed to open camera with pipeline: {pipeline}')
            return

        # 4. Setup Publishers
        self.raw_video_pub = self.create_publisher(Image, 'camera/raw_video', 1)
        self.result_img_pub = self.create_publisher(Image, 'camera/lane_detection_result', 1)
        self.bev_img_pub = self.create_publisher(Image, 'camera/bev', 1)
        self.mask_img_pub = self.create_publisher(Image, 'camera/bin_mask', 1)
        self.filter_img_pub = self.create_publisher(Image, 'camera/filter_bin_mask', 1)
        
        self.delta_pub = self.create_publisher(Int32, 'lane_detect/delta', 1)
        self.position_confidence_pub = self.create_publisher(Float64, 'lane_detect/position_confidence', 1)
        self.symmetry_confidence_pub = self.create_publisher(Float64, 'lane_detect/symmetry_confidence', 1)

        self.get_logger().info('lane_detection_node has started')

        # 5. Create Timer to trigger frame processing
        timer_period = 1.0 / fps
        self.timer = self.create_timer(timer_period, self.process_frame)

    def pub_img(self, publisher, img, encoding):
        """Helper to convert cv2 image to ROS Image msg and publish."""
        try:
            img_msg = self.bridge.cv2_to_imgmsg(img, encoding=encoding)
            img_msg.header.stamp = self.get_clock().now().to_msg()
            img_msg.header.frame_id = 'video_feed'
            publisher.publish(img_msg)
        except Exception as e:
            self.get_logger().error(f'Error publishing image: {e}')

    def process_frame(self):
        ret, frame = self.cap.read()

        if ret and frame is not None:
            # Convert BGRA to BGR if necessary
            if len(frame.shape) == 3 and frame.shape[2] == 4:
                frame = cv2.cvtColor(frame, cv2.COLOR_BGRA2BGR)
            
            # --- The custom processing step ---
            results = self.ld.step(frame)
            
            # Create and publish data messages based on results
            msg_delta = Int32(data=int(results.delta))
            msg_pos_conf = Float64(data=float(results.position_conf))
            msg_sym_conf = Float64(data=float(results.symmetry_conf))
            
            self.delta_pub.publish(msg_delta)
            self.position_confidence_pub.publish(msg_pos_conf)
            self.symmetry_confidence_pub.publish(msg_sym_conf)

            # Publish image streams
            self.get_logger().debug('pub images')
            self.pub_img(self.raw_video_pub, frame, 'bgr8')
            
            # Uncomment when your LaneDetector is ready:
            self.pub_img(self.result_img_pub, results.ld_img, 'bgr8')
            self.pub_img(self.bev_img_pub, results.bev, 'bgr8')
            self.pub_img(self.mask_img_pub, results.bin_mask, 'mono8')
            self.pub_img(self.filter_img_pub, results.filter_mask, 'mono8')

    def create_pipeline(self, cam_index, width, height, fps):
        """Constructs the GStreamer pipeline string for Jetson."""
        pipeline = (
            f"nvarguscamerasrc sensor-id={cam_index} "
            f"! video/x-raw(memory:NVMM),width=(int){width},height=(int){height},framerate=(fraction){fps}/1,format=NV12 "
            f"! nvvidconv "
            f"! video/x-raw,width=(int){width},height=(int){height},format=(string)BGRx "
            f"! videoconvert "
            f"! video/x-raw,format=BGR "
            f"! appsink sync=false"
)
        self.get_logger().info(f'Pipeline: {pipeline}')
        return pipeline

def main(args=None):
    rclpy.init(args=args)
    node = LaneDetectionNode()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()